### DISCUSSION TITLE
Problem with PWA but not SPA

### DISCUSSION THREAD

**Discussion by idc77** [keycloak/keycloak#29366] [Q&A]
Where to start.
 Well I'm desperate, Google Play Developer is threatening to delete my account, which I paid $25 for registration and demands I publish an app in 4 days.
But I have an issue now with Keycloak.
It's a recipe app, a cookbook at https://cooksbooks.de/ , very simple, created so my 89 year old mom can use it for her recipes.

I have 2 servers.
1 had keycloak on it via connect.icod.de
then other hosts the cooksbooks.de domain and content.
At first I had cooksbooks.de/auth/ reverse proxy to connect.icod.de and set the X-FRAME-OPTIONS to ALLOW-FROM to https://cooksbooks.de/
After a long evening and night I gave up, frustrated.
Nothing worked. I was always getting the `Timeout when waiting for 3rd party check iframe message.` from the keycloak-js package.

Then, next morning I had an idea, laying in bed.
"Let's move keycloak to Server2 and reverse proxy to localhost:4001"
Done and done.

Cooksbooks.de now has
```
server {
        listen 80;
        listen [::]:80;
        server_name cooksbooks.de;
        include /etc/nginx/acme.conf;
        error_log /var/log/nginx/cooksbooks.de_error.log;
        access_log /var/log/nginx/cooksbooks.de_access.log;
        location / {
                return 301 https://cooksbooks.de$request_uri;
        }
}

server {
    listen 443 ssl;
    listen [::]:443 ssl;
    http2 on;

    server_name cooksbooks.de;
    root /var/www/cooksbooks.de/frontend/spa;

    ssl_certificate /etc/letsencrypt/live/cooksbooks.de/fullchain.pem;
    ssl_certificate_key /etc/letsencrypt/live/cooksbooks.de/privkey.pem;

    error_log /var/log/nginx/cooksbooks.de_error.log;
    access_log /var/log/nginx/cooksbooks.de_access.log;

    location /auth/ {
        proxy_pass http://127.0.0.1:4001;
        proxy_set_header Host               $host;
        proxy_set_header X-Real-IP          $remote_addr;
        proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto  $scheme;
        proxy_buffer_size 64k;
        proxy_buffers 8 64k;
        proxy_busy_buffers_size 64k;
    }

    location /api/ {
       include cors.conf;
       proxy_pass REDACTED;
       proxy_read_timeout 30;
       proxy_connect_timeout 30;
       proxy_redirect off;
       proxy_set_header X-Real-IP          $remote_addr;
       proxy_set_header X-Forwarded-For    $proxy_add_x_forwarded_for;
       proxy_set_header X-Forwarded-Proto  $scheme;
    }

    location / {
        try_files $uri $uri/ /index.html;
    }

}
```

With SPA only I'm able to work around the issue with `Timeout when waiting for 3rd party check iframe message.`, which is displayed 3 times.
However with PWA active that doesn't work.
I opened an issue https://github.com/keycloak/keycloak/issues/29281 which was the initial issue with Chrome not liking what keycloak does.
But now it's, no matter if Chrome or Firefox, in PWA mode I'm getting 3x `error: "Timeout when waiting for 3rd party check iframe message."` on Chorme and 4x on Firefox.

"So don't use PWA", well I need PWA to be able to create a TWA and submit to the Google Play Store.

**Comment by codespearhead** [comment]
Allowing authentication via iframes will leave the users vulnerable to [Clickjacking](https://owasp.org/www-community/attacks/Clickjacking) and [Cross-Frame Scripting](https://owasp.org/www-community/attacks/Cross_Frame_Scripting) at best, which is probably the reason your Google Play Developer Account is currently in jeopardy: no platform should allow that.

Instead, you should use the Authorization Code Flow with PKCE for this type of client (see #28958).

Additionally, Make sure your production server that's behind a reverse proxy is properly configured (see #28279), and if you're deploying Keycloak with high-availability, which judging from what you've told us is an overkill, [you should properly configure infinispan](https://www.keycloak.org/high-availability/introduction).
